lib/cache.gi

```diff
@@ -0,0 +1,49 @@
+#############################################################################
+##
+#W  cache.gi                     GAP library                 Chris Jefferson
+##
+##
+#Y  Copyright (C) 2017 University of St Andrews, Scotland
+##
+##  This file defines various types of caching data structures
+##
+##  Note that this file is read very early in GAP's startup, so we cannot
+##  make this a method and use method dispatch, or other bits of nice
+##  GAP functionality.
+##
+
+InstallGlobalFunction(MemoizePosIntFunction,
+function(func, defaults...)
+    local boundvals, original, uniqueobj;
+
+    # This is an object which cannot exist anywhere else
+    uniqueobj := "";
+
+    if LEN_LIST(defaults) = 0 then
+        original := AtomicList([]);
+    else
+        original := AtomicList(defaults[1]);
+    fi;
+
+    boundvals := MakeWriteOnceAtomic(AtomicList(original));
+
+    InstallMethod(FlushCaches, [],
+        function()
+            boundvals := MakeWriteOnceAtomic(AtomicList(original));
+        end);
+
+    return function(val)
+        local v, boundcpy;
+        # Make copy in case cache gets flushed
+        boundcpy := boundvals;
```

Comment:
Nitpick, but: this might be slightly misleading, as we copy the *reference*, not the list itself. So perhaps: "# Make a copy of the reference to the list, in case another thread flushes the cache, which will cause boundvals to reference another list."